lib/Builder.js

```diff
@@ -394,18 +399,21 @@ CompiledNode.prototype.generateHashes = function (nodes, fnOriginNames, idProvid
   var hashBuilder = new NodeHashBuilder()
       .add(importantHashes, true)
       .add(argHashes)
-      .add(oid.hash(this.literalPromise))
+
+  if (this.literalPromise) {
+    hashBuilder.add(this._graph._oidProvider.getId(this.literalPromise))
```

Comment:
Lemme try to paraphrase to see if I understand what's happening here.

In Graph we have two {}'s. One maps a primitive literal to the promise that resolves it; the other maps a unique object ID to the promise that resolves that object.

When we generate the hash for a compiled node, we get the object ID of the *promise* and/or the resolver function, rather than for the object itself.

The main things you've changed are:

1. Using a WeakMap and a counter variable to assign unique IDs to non-primitives
2. Separating primitives from objects in the literal maps, because oid supported literals but the WeakMap strategy does not.

Is that right?

Reply:
@kylewm and I talked offline and he suggested a much simpler way to do this that I'll be doing.